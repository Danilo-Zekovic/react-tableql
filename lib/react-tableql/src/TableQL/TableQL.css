/* general styling */
body {
  font-family: 'Open Sans', sans-serif;
  line-height: 1.25;
}
