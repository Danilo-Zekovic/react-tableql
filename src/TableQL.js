import React, { useState } from 'react'
import PropTypes from 'prop-types'

import './index.css'

import Pagination from './Pagination'
import Table from './Table'
import ErrorBoundary from './ErrorBaundary'

const TableQL = ({
  pagination,
  debug,
  errorMessage,
  columns,
  styles,
  onRowClick,
  loading = false,
  error = '',
  data,
  sort,
}) => {
  const [currentPage, setCurrentPage] = useState(
    pagination && pagination.currentPage ? pagination.currentPage : 1,
  )

  const [displayData, setDisplayData] = useState()
  const [fullData, setFullData] = useState()

  // traverse data to find the array of objects and return it
  const traverseData = data => {
    log('Traverse data called.')
    for (let key in data) {
      if (Array.isArray(data)) {
        return data
      } else {
        return traverseData(data[key])
      }
    }
  }

  const getHeaderLabels = data => {
    log('Get header labels.')
    let labels = []
    for (let key in data) {
      // exception to eliminate meta fields
      if (!key.includes('__')) {
        labels.push(key)
      }
    }

    return labels
  }

  const onPageChanged = ({
    currentPage,
    totalPages,
    pageLimit,
    totalRecords,
  }) => {
    log(
      `On page changed: Current Page > ${currentPage}, Total Pages > ${totalPages}, Page Limit > ${pageLimit}, Total Records > ${totalRecords}`,
    )
    setCurrentPage(currentPage)

    // expose the values to parent if onPageChanged is passed as part of pagination
    if (pagination.onPageChanged) {
      pagination.onPageChanged(currentPage, totalPages, pageLimit, totalRecords)
    }
  }

  // when debug true log messages and data
  const log = (tag, load = '') => {
    if (debug) {
      console.log(tag, load)
    }
  }

  const onSort = column => {
    log('Sort by column: ', column)

    const property = typeof column === 'string' ? column : column.id
    let newOrder = [...fullData]

    if (column.sort && typeof column.sort === 'function') {
      newOrder = column.sort(newOrder, property)
    } else if ((column.sort && typeof column.sort === 'boolean') || sort) {
      newOrder.sort((a, b) => (a[property] > b[property] ? 1 : -1))
    }

    if (JSON.stringify(newOrder) === JSON.stringify(fullData)) {
      newOrder = newOrder.reverse()
    }

    setFullData(newOrder) // needed for pagination sort

    setDisplayData(newOrder)
  }

  log('Props: ', {
    pagination,
    debug,
    errorMessage,
    columns,
    styles,
    onRowClick,
  })

  if (loading) {
    log('Loading: ', loading)
    return <p>{`Loading...`}</p>
  }
  if (error) {
    log('Error: ', error)
    return (
      <p>{errorMessage || `Error while loading TableQL: ${error.message}`}</p>
    )
  }

  // TODO consider having something different when there is no data compared to empty array
  if (!data || data.length === 0) {
    log('No data found!')
    return <p>{`No data found!`}</p>
  }

  log('Data: ', data)

  // let displayData = traverseData(data)
  let traversedData = traverseData(data)

  if (!displayData) {
    setDisplayData(traversedData)
  }

  if (!fullData) {
    setFullData(traversedData)
  }

  let allData = traversedData
  let dataKeys = columns || getHeaderLabels(traversedData[0])

  let pageLimit
  if (pagination) {
    pageLimit = pagination.pageLimit || 10
    const offset =
      (currentPage - 1) *
      (pagination.pageLimit ? pagination.pageLimit : pageLimit)

    let dataToDisplay = displayData
      ? fullData.slice(offset).slice(0, pageLimit)
      : traversedData.slice(offset).slice(0, pageLimit)

    // it is ugly, but it will do it for now
    if (JSON.stringify(displayData) !== JSON.stringify(dataToDisplay)) {
      setDisplayData([...dataToDisplay])
    }
  }

  log('Data to be displayed (array): ', displayData)
  log('Data keys: ', dataKeys)

  // TODO probably bad idea not to display empty table
  if (!displayData || displayData.length == 0) {
    log('No data found!')
    return <p>{`No data found!`}</p>
  }
  return (
    <ErrorBoundary>
      <Table
        displayData={displayData}
        dataKeys={dataKeys}
        styles={styles}
        log={log}
        onRowClick={onRowClick}
        onSort={onSort}
        sort={sort}
      />

      {pagination && (
        <Pagination
          totalRecords={allData.length}
          pageLimit={pageLimit}
          pageNeighbors={pagination.pageNeighbors}
          selectedPage={currentPage}
          onPageChanged={returnedData => onPageChanged(returnedData)}
          styles={pagination.styles}
          log={log}
        />
      )}
    </ErrorBoundary>
  )
}

TableQL.propTypes = {
  data: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.object).isRequired,
    PropTypes.object,
  ]),
  loading: PropTypes.bool,
  error: PropTypes.object,
  errorMessage: PropTypes.string,
  columns: PropTypes.array,
  pagination: PropTypes.oneOfType([
    PropTypes.shape({
      pageLimit: PropTypes.number,
      pageNeighbors: PropTypes.number,
      currentPage: PropTypes.number,
      onPageChanged: PropTypes.func,
      styles: PropTypes.string,
    }),
    PropTypes.bool,
  ]),
  errorMessage: PropTypes.string,
  styles: PropTypes.shape({
    table: PropTypes.string,
    thead: PropTypes.string,
    theadTr: PropTypes.string,
    theadTh: PropTypes.string,
    tbody: PropTypes.string,
    tbodyTr: PropTypes.string,
    tbodyTd: PropTypes.string,
  }),
  onRowClick: PropTypes.func,
  sort: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
  debug: PropTypes.bool,
}

export default TableQL
