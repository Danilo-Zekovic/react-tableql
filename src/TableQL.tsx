import React, { useState, FC } from 'react'

import './index.css'

import Pagination, { PaginationData } from './Pagination'
import Table from './Table'
import ErrorBoundary from './ErrorBaundary'
import Loader from './components/Loader/Loader'

interface Props {
  data: object[]
  loading?: boolean
  error?: Error
  errorMessage?: string
  columns?: any[]
  pagination?:
    | {
        pageLimit?: number
        pageNeighbors?: number
        currentPage?: number
        onPageChanged?: (
          currentPage?: number,
          totalPages?: number,
          pageLimit?: number,
          totalRecords?: number,
        ) => void
        styles?: string
      }
    | boolean
  styles?: {
    table?: string
    thead?: string
    theadTr?: string
    theadTh?: string
    tbody?: string
    tbodyTr?: string
    tbodyTd?: string
  }
  onRowClick?: () => void
  sort?: boolean
  debug?: boolean
}

const TableQL: FC<Props> = ({
  pagination,
  debug,
  errorMessage,
  columns,
  styles,
  onRowClick,
  loading = false,
  error = new Error(),
  data,
  sort,
}) => {
  const [currentPage, setCurrentPage] = useState<number>(
    pagination && typeof pagination !== 'boolean' && pagination.currentPage
      ? pagination.currentPage
      : 1,
  )

  const [displayData, setDisplayData] = useState<any>([])
  const [fullData, setFullData] = useState<any>([])

  // traverse data to find the array of objects and return it
  const traverseData = (data: any) => {
    log('Traverse data called.')
    for (let key in data) {
      if (Array.isArray(data)) {
        return data
      } else {
        return traverseData(data[key])
      }
    }
  }

  const getHeaderLabels = (data: any) => {
    log('Get header labels.')
    let labels = [] as string[]
    for (let key in data) {
      // exception to eliminate meta fields
      if (!key.includes('__')) {
        labels.push(key)
      }
    }

    return labels
  }

  const onPageChanged = ({
    currentPage,
    totalPages,
    pageLimit,
    totalRecords,
  }: PaginationData) => {
    log(
      `On page changed: Current Page > ${currentPage}, Total Pages > ${totalPages}, Page Limit > ${pageLimit}, Total Records > ${totalRecords}`,
    )
    if (currentPage) setCurrentPage(currentPage)

    // expose the values to parent if onPageChanged is passed as part of pagination
    if (typeof pagination === 'object' && pagination.onPageChanged) {
      pagination.onPageChanged(currentPage, totalPages, pageLimit, totalRecords)
    }
  }

  // when debug true log messages and data
  const log = (tag: string, load = '' as any) => {
    if (debug) {
      console.log(tag, load)
    }
  }

  const onSort = (column: any) => {
    log('Sort by column: ', column)

    const property = typeof column === 'string' ? column : column.id
    let newOrder = [...fullData]

    if (column.sort && typeof column.sort === 'function') {
      newOrder = column.sort(newOrder, property)
    } else if ((column.sort && typeof column.sort === 'boolean') || sort) {
      newOrder.sort((a, b) => (a[property] > b[property] ? 1 : -1))
    }

    if (JSON.stringify(newOrder) === JSON.stringify(fullData)) {
      newOrder = newOrder.reverse()
    }

    setFullData(newOrder) // needed for pagination sort

    setDisplayData(newOrder)
  }

  log('Props: ', {
    pagination,
    debug,
    errorMessage,
    columns,
    styles,
    onRowClick,
  })

  if (loading) {
    log('Loading: ', loading)
    return <Loader />
  }
  if (error) {
    log('Error: ', error)
    return (
      <p>{errorMessage || `Error while loading TableQL: ${error.message}`}</p>
    )
  }

  // TODO consider having something different when there is no data compared to empty array
  if (!data || data.length === 0) {
    log('No data found!')
    return <p>{`No data found!`}</p>
  }

  log('Data: ', data)

  // let displayData = traverseData(data)
  let traversedData = traverseData(data)

  if (!displayData) {
    setDisplayData(traversedData)
  }

  if (!fullData) {
    setFullData(traversedData)
  }

  let allData = traversedData
  let dataKeys = columns || getHeaderLabels(traversedData[0])

  let pageLimit = 10
  if (pagination) {
    pageLimit =
      typeof pagination !== 'boolean' && pagination.pageLimit
        ? pagination.pageLimit
        : 10
    const offset =
      (currentPage - 1) *
      (typeof pagination !== 'boolean' && pagination.pageLimit
        ? pagination.pageLimit
        : pageLimit)

    let dataToDisplay = displayData
      ? fullData.slice(offset).slice(0, pageLimit)
      : traversedData.slice(offset).slice(0, pageLimit)

    // it is ugly, but it will do it for now
    if (JSON.stringify(displayData) !== JSON.stringify(dataToDisplay)) {
      setDisplayData([...dataToDisplay])
    }
  }

  log('Data to be displayed (array): ', displayData)
  log('Data keys: ', dataKeys)

  // TODO probably bad idea not to display empty table
  if (!displayData || displayData.length == 0) {
    log('No data found!')
    return <p>{`No data found!`}</p>
  }
  return (
    <ErrorBoundary>
      <Table
        displayData={displayData}
        // @ts-ignore
        dataKeys={dataKeys}
        styles={styles}
        log={log}
        onRowClick={onRowClick}
        onSort={onSort}
        sort={sort}
      />

      {pagination && (
        <Pagination
          totalRecords={allData.length}
          pageLimit={pageLimit}
          pageNeighbors={
            typeof pagination !== 'boolean'
              ? pagination.pageNeighbors
              : undefined
          }
          selectedPage={currentPage}
          onPageChanged={(returnedData: PaginationData) =>
            onPageChanged(returnedData)
          }
          styles={
            typeof pagination !== 'boolean' ? pagination.styles : undefined
          }
          log={log}
        />
      )}
    </ErrorBoundary>
  )
}

export default TableQL
