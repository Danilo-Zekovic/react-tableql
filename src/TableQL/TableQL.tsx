import React, { useState, FC, useEffect, useCallback, useMemo } from 'react'

import '../index.css'

import Pagination, { PaginationData } from '../components/Pagination/Pagination'
import Table, { Styles, ColumnConfig } from '../components/Table/Table'
import ErrorBoundary from '../components/ErrorBoundary/ErrorBoundary'
import Loader from '../components/Loader/Loader'
import { traverseObjectForArray } from '../utils'

export interface PaginationConfig {
  pageLimit?: number
  pageNeighbors?: number
  currentPage?: number
  onPageChanged?: (paginationData: PaginationData) => void
  styles?: string
}

export interface FailSafe {
  reason: string
}

export interface Props {
  data: object[] | object | []
  loading?: boolean
  error?: Error
  errorMessage?: string
  columns?: (string | ColumnConfig)[]
  pagination?: PaginationConfig | boolean
  styles?: Styles
  onRowClick?: (data: { [key: string]: unknown }) => void
  sort?: boolean
  debug?: boolean
  onEmpty?: FC<FailSafe>
}

const TableQL: FC<Props> = ({
  pagination,
  debug,
  errorMessage,
  columns,
  styles,
  onRowClick,
  loading = false,
  error,
  data,
  sort,
  onEmpty,
}) => {
  const [currentPage, setCurrentPage] = useState<number>(
    pagination && typeof pagination !== 'boolean' && pagination.currentPage
      ? pagination.currentPage
      : 1,
  )

  // Data that is visible to user
  const [displayData, setDisplayData] = useState<any[]>()
  // All data that is important when pagination true, because in that case not all of the data is visible
  const [fullData, setFullData] = useState<any[]>([])

  const [dataKeys, setDataKeys] = useState<(string | ColumnConfig)[]>([])

  const isEmpty = useMemo(
    () =>
      (!displayData || displayData.length === 0) &&
      !loading &&
      displayData !== undefined,
    [displayData, loading],
  )

  useEffect(() => {
    log('Data or columns changed: ', { data, columns })
    const traversedData = data ? traverseObjectForArray(data) : []

    setDisplayData(traversedData)
    setFullData(traversedData)
    setDataKeys(columns || getHeaderLabels(traversedData[0]))
  }, [data, columns])

  const getHeaderLabels = (data: any) => {
    log('Get header labels.')
    let labels = [] as string[]
    for (let key in data) {
      // exception to eliminate meta fields
      if (!key.includes('__')) {
        labels.push(key)
      }
    }

    return labels
  }

  const onPageChanged = ({
    currentPage,
    totalPages,
    pageLimit,
    totalRecords,
  }: PaginationData) => {
    log(
      `On page changed: Current Page > ${currentPage}, Total Pages > ${totalPages}, Page Limit > ${pageLimit}, Total Records > ${totalRecords}`,
    )
    if (currentPage) setCurrentPage(currentPage)

    // expose the values to parent if onPageChanged is passed as part of pagination
    if (typeof pagination === 'object' && pagination.onPageChanged) {
      pagination.onPageChanged({
        currentPage,
        totalPages,
        pageLimit,
        totalRecords,
      })
    }
  }

  // when debug true log messages and data
  const log = (tag: string, load = '' as any) => {
    if (debug) {
      console.log(tag, load)
    }
  }

  const onSort = (column: string | ColumnConfig) => {
    log('Sort by column: ', column)

    const property = typeof column === 'string' ? column : column.id
    let newOrder = fullData ? ([...fullData] as any) : []

    if (
      typeof column !== 'string' &&
      column.sort &&
      typeof column.sort === 'function'
    ) {
      newOrder = column.sort(newOrder, property)
    } else if (
      (typeof column !== 'string' &&
        column.sort &&
        typeof column.sort === 'boolean') ||
      sort
    ) {
      newOrder.sort((a: unknown[], b: unknown[]) =>
        a[property] > b[property] ? 1 : -1,
      )
    }

    if (JSON.stringify(newOrder) === JSON.stringify(fullData)) {
      newOrder = newOrder.reverse()
    }

    setFullData(newOrder) // needed for pagination sort

    setDisplayData(newOrder)
  }

  if (!(window !== undefined) || !(document !== undefined)) {
    log('Window or document are not defined.')
    return null
  }

  log('Props: ', {
    pagination,
    debug,
    errorMessage,
    columns,
    styles,
    onRowClick,
  })

  if (loading || displayData === undefined || data === undefined) {
    log('Loading: ', loading)
    return <Loader />
  }
  if (error) {
    log('Error: ', error)
    return (
      <p>{errorMessage || `Error while loading TableQL: ${error.message}`}</p>
    )
  }
  // ======================
  // TODO consider having something different when there is no data compared to empty array
  if (Array.isArray(data) && data.length === 0) {
    log('No data found!', data.length)
    if (onEmpty) {
      return onEmpty({ reason: 'No data found!' })
    }
    return <p>{`No data found!`}</p>
  }

  log('Data: ', data)

  let pageLimit = 10
  if (pagination && fullData && displayData) {
    pageLimit =
      typeof pagination !== 'boolean' && pagination.pageLimit
        ? pagination.pageLimit
        : 10
    const offset =
      (currentPage - 1) *
      (typeof pagination !== 'boolean' && pagination.pageLimit
        ? pagination.pageLimit
        : pageLimit)

    if (!displayData || displayData.length === 0) {
      log('No data found to paginate!', displayData)
      if (onEmpty) {
        return onEmpty({ reason: 'No data found to paginate!' })
      }
      return <p>{`No data found!`}</p>
    }
    // let dataToDisplay =
    //   displayData && fullData
    //     ? fullData.slice(offset).slice(0, pageLimit)
    //     : traversedData.slice(offset).slice(0, pageLimit)

    const dataToDisplay = fullData.slice(offset).slice(0, pageLimit)

    // it is ugly, but it will do it for now
    if (JSON.stringify(displayData) !== JSON.stringify(dataToDisplay)) {
      setDisplayData([...dataToDisplay])
    }
  }

  log('Data to be displayed (array): ', displayData)
  log('Data keys: ', dataKeys)

  // TODO probably bad idea not to display empty table
  // if ((!displayData || displayData.length === 0) && !loading && displayData !== undefined) {
  if (isEmpty) {
    log('No data found! I am dumb shit', displayData)
    if (onEmpty) {
      return onEmpty({ reason: 'No data found!' })
    }
    return <p>{`No data found!`}</p>
  }

  return (
    <ErrorBoundary>
      <Table
        displayData={displayData}
        dataKeys={dataKeys}
        styles={styles}
        log={log}
        onRowClick={onRowClick}
        onSort={onSort}
        sort={sort}
      />

      {pagination && (
        <Pagination
          totalRecords={fullData.length}
          pageLimit={pageLimit}
          pageNeighbors={
            typeof pagination !== 'boolean'
              ? pagination.pageNeighbors
              : undefined
          }
          selectedPage={currentPage}
          onPageChanged={(returnedData: PaginationData) =>
            onPageChanged(returnedData)
          }
          styles={
            typeof pagination !== 'boolean' ? pagination.styles : undefined
          }
          log={log}
        />
      )}
    </ErrorBoundary>
  )
}

export default TableQL
