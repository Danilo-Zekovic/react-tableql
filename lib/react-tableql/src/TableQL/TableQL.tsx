import React, { FC } from 'react'

import { TableProvider } from '../TableProvider'
import { Table } from '../components/Table'

import './TableQL.css'

const TableQL: FC = () => {
  return <Table />
}

const TableQLWithProvider: FC = ({ ...props }) => (
  <TableProvider>
    <TableQL {...props} />
  </TableProvider>
)

export default TableQLWithProvider
