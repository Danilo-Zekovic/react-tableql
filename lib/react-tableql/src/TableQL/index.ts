import TableQL from './TableQL'

export { TableQL }
