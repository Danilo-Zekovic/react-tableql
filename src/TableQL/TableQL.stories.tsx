import React from 'react'

import TableQL from './TableQL'
import { FILMS } from '../../__mocks__/data'
import { STORYBOOK_SECTIONS } from '../../stories/constants'

export default {
  title: `${STORYBOOK_SECTIONS.tableQl}/Basic`,
  component: TableQL,
}

export const Default = () => <TableQL data={FILMS} />

export const NoData = () => (
  <TableQL
    data={{ data: [] }}
    columns={[
      { id: 'foo', label: 'Foo', sort: true },
      { id: 'bar', label: 'Bar', sort: false },
    ]}
    pagination
  />
)

export const OnEmpty = () => (
  <TableQL
    data={{ data: [] }}
    columns={[
      { id: 'foo', label: 'Foo', sort: true },
      { id: 'bar', label: 'Bar', sort: false },
    ]}
    onEmpty={({ reason }) => <h3>You don't have data yet. Reason {reason}</h3>}
  />
)

export const IncorectData = () => <TableQL data={{ data: 'asd' }} />
