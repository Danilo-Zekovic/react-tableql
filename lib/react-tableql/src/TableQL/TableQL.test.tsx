import React from 'react'

import { render } from '@testing-library/react'

import TableQL from './TableQL'

describe('<TableQL>', () => {
  it('default snapshot', () => {
    const { container } = render(<TableQL />)

    expect(container.firstChild).toMatchSnapshot()
  })
})
