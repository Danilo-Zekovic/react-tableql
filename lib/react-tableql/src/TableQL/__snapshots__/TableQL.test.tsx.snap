// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableQL> default snapshot 1`] = `
<table
  class="TableQL"
>
  <caption
    class="TableQL-caption"
  >
    Statement Summary
  </caption>
  <thead
    class="TableQL-thead"
  >
    <tr
      class="TableQL-tr"
    >
      <th
        class="TableQL-th"
        scope="col"
      >
        Account
      </th>
      <th
        class="TableQL-th"
        scope="col"
      >
        Due Date
      </th>
      <th
        class="TableQL-th"
        scope="col"
      >
        Amount
      </th>
      <th
        class="TableQL-th"
        scope="col"
      >
        Period
      </th>
    </tr>
  </thead>
  <tbody
    class="TableQL-tbody"
  >
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="Account"
      >
        Visa - 3412
      </td>
      <td
        class="TableQL-td"
        data-label="Due Date"
      >
        04/01/2016
      </td>
      <td
        class="TableQL-td"
        data-label="Amount"
      >
        $1,190
      </td>
      <td
        class="TableQL-td"
        data-label="Period"
      >
        03/01/2016 - 03/31/2016
      </td>
    </tr>
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="Account"
        scope="row"
      >
        Visa - 6076
      </td>
      <td
        class="TableQL-td"
        data-label="Due Date"
      >
        03/01/2016
      </td>
      <td
        class="TableQL-td"
        data-label="Amount"
      >
        $2,443
      </td>
      <td
        class="TableQL-td"
        data-label="Period"
      >
        02/01/2016 - 02/29/2016
      </td>
    </tr>
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="Account"
        scope="row"
      >
        Corporate AMEX
      </td>
      <td
        class="TableQL-td"
        data-label="Due Date"
      >
        03/01/2016
      </td>
      <td
        class="TableQL-td"
        data-label="Amount"
      >
        $1,181
      </td>
      <td
        class="TableQL-td"
        data-label="Period"
      >
        02/01/2016 - 02/29/2016
      </td>
    </tr>
    <tr
      class="TableQL-tr"
    >
      <td
        class="TableQL-td"
        data-label="Acount"
        scope="row"
      >
        Visa - 3412
      </td>
      <td
        class="TableQL-td"
        data-label="Due Date"
      >
        02/01/2016
      </td>
      <td
        class="TableQL-td"
        data-label="Amount"
      >
        $842
      </td>
      <td
        class="TableQL-td"
        data-label="Period"
      >
        01/01/2016 - 01/31/2016
      </td>
    </tr>
  </tbody>
</table>
`;
